Source/Core/DolphinQt2/Resources.cpp

```diff
@@ -2,46 +2,74 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include <QGuiApplication>
+#include <QIcon>
+#include <QScreen>
 #include <QStringList>
 
 #include "Common/CommonPaths.h"
 #include "Common/FileUtil.h"
 #include "DolphinQt2/Resources.h"
+#include "DolphinQt2/Settings.h"
 
 QList<QPixmap> Resources::m_platforms;
 QList<QPixmap> Resources::m_countries;
 QList<QPixmap> Resources::m_ratings;
 QList<QPixmap> Resources::m_misc;
 
+QString Resources::GetPreferredPixmapScale(int max)
+{
+  const auto dpr = QGuiApplication::primaryScreen()->devicePixelRatio();
+  const auto scale = std::min(dpr <= 1 ? 1 : (dpr <= 2) ? 2 : 4, max);
+
+  if (scale > 1)
+    return QStringLiteral("@%1x").arg(scale);
+  else
+    return QStringLiteral("");
+}
+
+QPixmap Resources::GetScaledPixmap(const std::string& name, bool is_themed)
+{
+  QPixmap pixmap;
+  for (int scale : {4, 2, 1})
```

Comment:
If devicePixelRatio is less than 4 (which it will be everywhere right now) or less than 2, then this is needlessly repeating the same code.

Something unrolled like this seems simpler to me:

```c++
const float screen_ratio = QGuiApplication::primaryScreen()->devicePixelRatio();
std::string dir = is_themed ? Settings().GetThemeDir() : Settings().GetResourcesDir();

if (screen_ratio > 2) {
  QPixmap pixmap(dir + name + "@4x.png");
  pixmap.setDevicePixelRatio(4.0);
  if (!pixmap.isNull()) return pixmap;
}
if (screen_ratio > 1) {
  QPixmap pixmap(dir + name + "@2x.png");
  if (!pixmap.isNull()) return pixmap;
}
QPixmap pixmap(dir + name + "png");
return pixmap;
```